client/core/core.go

```diff
@@ -1923,6 +1911,68 @@ func (c *Core) WalletState(assetID uint32) *WalletState {
 	return wallet.state()
 }
 
+func (c *Core) walletCheckAndNotify(w *xcWallet) bool {
+	synced, progress, err := w.SyncStatus()
+	if err != nil {
+		c.log.Errorf("Unable to get wallet/node sync status for %s: %v",
+			unbip(w.AssetID), err)
+		return false
+	}
+	w.mtx.Lock()
+	w.synced = synced
+	w.syncProgress = progress
+	w.mtx.Unlock()
+	c.notify(newWalletStateNote(w.state()))
+	if synced {
+		c.updateWalletBalance(w)
+		c.log.Infof("Wallet synced for asset %s", unbip(w.AssetID))
+	}
+	return synced
+}
+
+// RescanWallet will issue a rescan command to the wallet if supported by the
+// wallet implementation. It is up to the underlying wallet backend if and how
+// to implement this functionality. It may be asynchronous. Core will emit
+// wallet state notifications until the rescan is complete. WARNING: It is
+// ill-advised to initiate a wallet rescan with active orders unless as a last
+// ditch effort to get the wallet to recognize a transaction needed to complete
+// a swap.
+func (c *Core) RescanWallet(assetID uint32) error {
+	wallet, err := c.connectedWallet(assetID)
+	if err != nil {
+		return fmt.Errorf("OpenWallet: wallet not found for %d -> %s: %w",
+			assetID, unbip(assetID), err)
+	}
+
+	// Begin potentially asynchronous wallet rescan operation.
+	if err = wallet.Rescan(c.ctx); err != nil {
+		return err
+	}
+
+	if c.walletCheckAndNotify(wallet) {
+		return nil // sync done
+	}
+
+	// Resync still running. Launch a status update goroutine.
+	go func() {
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				if c.walletCheckAndNotify(wallet) {
+					return
+				}
+
+			case <-c.ctx.Done():
+				return
+			}
+		}
+	}()
```

Comment:
Maybe fine, but I want to point out that both the wallet's `rescanWalletAsync` and core's `updateWalletBalance` have db writes and nothing is waiting for this goroutine to finish. Maybe golang will handle this fine, but what if it just exits the app in the middle of a db write? imo app shutdown should wait for this to finish.

Reply:
Good point.  I'll see about making it a supervised task that holds up shutdown.